src/one/nio/serial/JsonReader.java

```diff
@@ -193,23 +199,33 @@ public Object readNull() throws IOException {
     }
 
     public String readString() throws IOException {
-        StringBuilder sb = new StringBuilder();
-        expect('\"', "Expected string");
-        while (next >= 0 && next != '\"') {
-            int b = read();
-            if ((b & 0x80) == 0) {
-                sb.append(b == '\\' ? readEscapeChar() : (char) b);
-            } else if ((b & 0xe0) == 0xc0) {
-                sb.append((char) ((b & 0x1f) << 6 | (read() & 0x3f)));
-            } else if ((b & 0xf0) == 0xe0) {
-                sb.append((char) ((b & 0x0f) << 12 | (read() & 0x3f) << 6 | (read() & 0x3f)));
-            } else {
-                int v = (b & 0x07) << 18 | (read() & 0x3f) << 12 | (read() & 0x3f) << 6 | (read() & 0x3f);
-                sb.append((char) (0xd800 | (v - 0x10000) >>> 10)).append((char) (0xdc00 | (v & 0x3ff)));
-            }
+        switch (next) {
+            case '\"':
+                StringBuilder sb = new StringBuilder();
+                read();
+                while (next >= 0 && next != '\"') {
+                    int b = read();
+                    if ((b & 0x80) == 0) {
+                        sb.append(b == '\\' ? readEscapeChar() : (char) b);
+                    } else if ((b & 0xe0) == 0xc0) {
+                        sb.append((char) ((b & 0x1f) << 6 | (read() & 0x3f)));
+                    } else if ((b & 0xf0) == 0xe0) {
+                        sb.append((char) ((b & 0x0f) << 12 | (read() & 0x3f) << 6 | (read() & 0x3f)));
+                    } else {
+                        int v = (b & 0x07) << 18 | (read() & 0x3f) << 12 | (read() & 0x3f) << 6 | (read() & 0x3f);
+                        sb.append((char) (0xd800 | (v - 0x10000) >>> 10)).append((char) (0xdc00 | (v & 0x3ff)));
+                    }
+                }
+                expect('\"', "Unexpected end of string");
+                return sb.toString();
+            case 'f':
+            case 'F':
+            case 't':
+            case 'T':
+                return "" + readBoolean();
+            default:
+                return readNumber();
```

Comment:
Error reporting might be misleading due to fallback to number parsing.
What do you think about replacing `default` case with supported chars for the start of a number and an explicit error case about unexpected input:
```java
case '-':
case '0':
case '1':
...
case '9':
    return readNumber();
default:
    throw exception("Unexpected start of string");
```